### PR TITLE
Allow using MariaDb instead of MySQL for tests

### PR DESCRIPTION
**Pull request by sgiehl**
### Description:

As we officially support MariaDb it might be useful to run some of our tests with MariaDb instead of MySQL to discover possible incompatibilities earlier. 

This PR will allow a test action to decide which database to use.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -58,6 +58,13 @@ inputs:
     description: "If true a MySQL 5.7 will be set up"
     required: false
     default: true
+  mysql-schema:
```
Comment (michalkleiner) [inline action.yml:61]: I'd go with a generic db-provider here as it will allow us to run different database engines as well, not being forks or feature-matching engines with MySQL.
```suggestion
  db-provider:
```

File: action.yml [hunk 2]
```diff
@@ -58,6 +58,13 @@ inputs:
...
+    description: "MySQL database schema to use. Can be Mysql or Mariadb"
+    options:
+      - Mysql
+      - Mariadb
+    required: false
+    default: 'Mysql'
```
Comment (michalkleiner) [inline action.yml:67]: I'd go with a generic db-provider here as it will allow us to run different database engines as well, not being forks or feature-matching engines with MySQL.

Then I would have the values all lowercase as then we don't need to worry about correct capitalisation of the names.

```suggestion
  db-provider:
    description: "Database provider to use. Can be 'mysql' or 'mariadb'."
    options:
      - mysql
      - mariadb
    required: false
    default: 'mysql'
```

File: action.yml [hunk 3]
```diff
@@ -89,13 +96,21 @@ runs:
   steps:
 
     - name: start mysql services
-      if: inputs.mysql-service == 'true'
+      if: inputs.mysql-service == 'true' && inputs.mysql-schema == 'Mysql'
```
Comment (michalkleiner) [inline action.yml:99]:
```suggestion
      if: inputs.mysql-service == 'true' && inputs.db-provider == 'mysql'
```

File: action.yml [hunk 4]
```diff
@@ -89,13 +96,21 @@ runs:
...
       shell: bash
       run: |
         docker run -d --tmpfs /var/lib/mysql:rw --tmpfs /bitnami/mysql/data:rw -v ${{ github.action_path }}/artifacts/my.cnf:/opt/bitnami/mysql/conf/my_custom.cnf:ro --name mysql -p 3306:3306 -e ALLOW_EMPTY_PASSWORD=yes -e MYSQL_DATABASE=matomo_tests bitnami/mysql:5.7
         sleep 10
         mysql -h127.0.0.1 -uroot -e "SET GLOBAL sql_mode = 'NO_ENGINE_SUBSTITUTION,STRICT_TRANS_TABLES';SET GLOBAL wait_timeout = 36000;SET GLOBAL max_allowed_packet = 134209536;"
 
+    - name: start mariadb services
+      if: inputs.mysql-service == 'true' && inputs.mysql-schema == 'Mariadb'
```
Comment (michalkleiner) [inline action.yml:107]:
```suggestion
      if: inputs.mysql-service == 'true' && inputs.db-provider == 'mariadb'
```

File: action.yml [hunk 5]
```diff
@@ -179,6 +194,7 @@ runs:
         TEST_SUITE: ${{ inputs.test-type }}
         MATOMO_TEST_TARGET: ${{ inputs.matomo-test-branch }}
         MYSQL_ADAPTER: ${{ inputs.mysql-driver }}
+        MYSQL_SCHEMA: ${{ inputs.mysql-schema }}
```
Comment (michalkleiner) [inline action.yml:197]:
```suggestion
        DB_PROVIDER: ${{ inputs.db-provider }}
```

File: action.yml [hunk 6]
```diff
@@ -222,6 +238,7 @@ runs:
         PHPUNIT_EXTRA_OPTIONS: ${{ inputs.phpunit-test-options }}
         GITHUB_BRANCH: ${{ github.head_ref || github.ref_name }}
         MYSQL_ADAPTER: ${{ inputs.mysql-driver }}
+        MYSQL_SCHEMA: ${{ inputs.mysql-schema }}
```
Comment (michalkleiner) [inline action.yml:241]:
```suggestion
        DB_PROVIDER: ${{ inputs.db-provider }}
```

File: artifacts/config.ini.github.php [hunk 7]
```diff
@@ -9,6 +9,7 @@
 ; we need to use piwik_tests here, as otherwise tests will fail for Matomo 4 versions before the rename
 dbname = piwik_tests
 adapter = PDO_MYSQL
+schema = Mysql
```
Comment (michalkleiner) [inline artifacts/config.ini.github.php:12]:
```suggestion
provider = mysql
```

File: artifacts/config.ini.github.php [hunk 8]
```diff
@@ -23,6 +24,7 @@
 password =
 dbname = piwik_tests
 adapter = PDO_MYSQL
+schema = Mysql
```
Comment (michalkleiner) [inline artifacts/config.ini.github.php:27]:
```suggestion
provider = mysql
```

File: scripts/bash/prepare.sh [hunk 9]
```diff
@@ -43,7 +43,7 @@ if [ "$PHP_VERSION" = "8.1" ] || [ "$PHP_VERSION" = "8.2" ] || [ "$PHP_VERSION"
 fi
 
 # setup config
-sed "s/PDO_MYSQL/$MYSQL_ADAPTER/g" $ACTION_PATH/artifacts/config.ini.github.php >config/config.ini.php
+sed "s/PDO_MYSQL/$MYSQL_ADAPTER/g; s/schema = Mysql/schema = $MYSQL_SCHEMA/g" $ACTION_PATH/artifacts/config.ini.github.php >config/config.ini.php
```
Comment (michalkleiner) [inline scripts/bash/prepare.sh:46]:
```suggestion
sed "s/PDO_MYSQL/$MYSQL_ADAPTER/g; s/provider = mysql/provider = $DB_PROVIDER/g" $ACTION_PATH/artifacts/config.ini.github.php >config/config.ini.php
```

File: action.yml (hunk 2 above)
Comment (michalkleiner) [reply action.yml:67]: Alternatively db-engine or mysql-engine, but I don't think schema is correct here, that has a different meaning imho.

**Comment by michalkleiner** [comment]
I found there's a PR using this https://github.com/matomo-org/matomo/pull/18371 where the code suggests Mysql is a schema and Mariadb extends from that and it's also called a schema. I think that's an unfortunate legacy that we need to carry forward at this stage. We can possibly open that to other database engines in Matomo 6, if we wanted to. From that perspective, I'd have to be ok with us calling the different db provider a schema :-/

File: artifacts/config.ini.github.php (hunk 8 above)
Comment (sgiehl) [reply artifacts/config.ini.github.php:27]: We can't change that. that would require a rename of that config setting in Matomo.

**Comment by sgiehl** [comment]
I agree that schema is not really the correct wording in general. I'll update it here to db-provider where possible. But we need to keep it in some places so it's compatible with Matomo. Also the names need to be defined how they can be used in Matomo's config. We can rework that if we ever refactor/change that in Matomo.

**Review by michalkleiner — APPROVED** [review]

**Comment by michalkleiner** [comment]
Thanks for considering and implementing the feedback @sgiehl, appreciate that. I'm happy with this change ✅.
